### PR TITLE
Feature/share account team organization

### PR DESCRIPTION
**Pull request by Lolincolc**
This PR adds an enpoint to share an account with another user whether it exists or not. It also adds a new table in the DB.

New endpoint : /user/invite
The endpoint takes three parameters : 

- Email : the mail of the invited user
- AccountId : the AWS account ID (from table aws_account) that the user wants to share access to
- PermissionLevel : the permission level (0 : admin access, 1 : standard user access, 2 : read-only user)

### How this is working :
- If the user does not exist, a new account is created with a temporary password randomly generated as a placeholder in the DB. A Forgotten password token is also generated allowing the user to choose his password before first login. Then, a new entry in table shared_account is inserted (see shared_account detail section).
- If the user already exist, a new entry is inserted in table shared_account

Users then receive an email that inform them that a new account has been added to their account. If the user didn't exist previously, the user is invited to create his account with the reset password link provided in the mail.

### New shared_account table details : 
- id: the id of the new entry
- account_id: the AWS account id that is being shared
- owner_id: user id of the user who is inviting someone else
- user_id : user id of the invited user
- user_permission: permission level
- account_status: 0 = Invited user has never logged in, 1 = Invited user has logged at least once

### CODE REVIEW

File: db/migration/0011_account_sharing.sql [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+CREATE TABLE shared_account (
+	id                     INTEGER      NOT NULL AUTO_INCREMENT,
+	account_id             INTEGER      NOT NULL,
+	owner_id               INTEGER      NOT NULL,
+	user_id                INTEGER      NOT NULL,
+	user_permission        TINYINT(2)   NOT NULL DEFAULT 0,
+	account_status         TINYINT(1)   NOT NULL DEFAULT 0,
+	CONSTRAINT PRIMARY KEY (id),
+  CONSTRAINT foreign_aws_account FOREIGN KEY (account_id) REFERENCES aws_account(id) ON DELETE CASCADE,
```
Comment (simonmeyerrr) [inline db/migration/0011_account_sharing.sql]: bad alignment on the 3 lines

File: db/schema.sql [hunk 2]
```diff
@@ -276,3 +277,30 @@ CREATE TABLE emailed_anomaly (
...
+
+CREATE TABLE shared_account (
+	id                     INTEGER      NOT NULL AUTO_INCREMENT,
+	account_id             INTEGER      NOT NULL,
+	owner_id               INTEGER      NOT NULL,
+	user_id                INTEGER      NOT NULL,
+	user_permission        TINYINT(2)   NOT NULL DEFAULT 0,
+	account_status         TINYINT(1)   NOT NULL DEFAULT 0,
+	CONSTRAINT PRIMARY KEY (id),
+  CONSTRAINT foreign_aws_account FOREIGN KEY (account_id) REFERENCES aws_account(id) ON DELETE CASCADE,
```
Comment (simonmeyerrr) [inline db/schema.sql]: same

File: users/shared_account.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+	return false,nil
+}
+
+// addAccountToGuest adds an entry in shared_account table allowing a user
+// to share an access to all or part of his account
+func addAccountToGuest(ctx context.Context, db *sql.Tx, accountId int, permissionLevel int, guestId int, ownerId int) (err error) {
+	const sqlstr = `INSERT INTO shared_account(
+			account_id, owner_id, user_id, user_permission, account_status
+		) VALUES (?, ?, ?, ?, ?)`
+	res, err := db.Exec(sqlstr, accountId, ownerId, guestId, permissionLevel, 0)
```
Comment (simonmeyerrr) [inline users/shared_account.go]: use `_, err := ...` instand of `res, err := ...    /   _ = res`

File: users/shared_account.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+	_ = res
+	if err != nil {
```
Comment (simonmeyerrr) [inline users/shared_account.go]: this condition is useless, you can just make `return err` in both case

File: users/shared_account.go [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+// createAccountForGuest creates an account for invited user who do not already own an account
+func createAccountForGuest(ctx context.Context, db *sql.Tx, userMail string, accountId int, permissionLevel int, user User) (newUserId int, err error) {
+	logger := jsonlog.LoggerFromContextOrDefault(ctx)
+	tempPassword := uuid.NewV1().String()
+	usr, err := CreateUserWithPassword(ctx, db, userMail, tempPassword, "")
+	if err == nil {
+		err = addAccountToGuest(ctx, db, accountId, permissionLevel, usr.Id, user.Id)
+		if err != nil {
+			logger.Warning("Error occured while adding account to an newly created user.", err)
```
Comment (simonmeyerrr) [inline users/shared_account.go]: Why Warning and not Error ? use `err.Error`

File: users/shared_account.go [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+			return 0, err
+		}
+	} else {
+		logger.Warning("Error occured while creating an automatic new account.", err)
```
Comment (simonmeyerrr) [inline users/shared_account.go]: same

**Review by simonmeyerrr — CHANGES_REQUESTED** [review]
There are changes to be made.
You don't need to put a variable name from the return prototype of a function (example: you always put `err error` when you could just put `error`)
the  `inviteUserWithValidBody` function is not well readable, is there not a way to improve it?

File: db/migration/0011_account_sharing.sql [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+--
+--   Unless required by applicable law or agreed to in writing, software
+--   distributed under the License is distributed on an "AS IS" BASIS,
+--   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+--   See the License for the specific language governing permissions and
+--   limitations under the License.
+
+CREATE TABLE shared_account (
+  id                     INTEGER      NOT NULL AUTO_INCREMENT,
+  account_id             INTEGER      NOT NULL,
```
Comment (thibautcornolti) [inline db/migration/0011_account_sharing.sql]: `owner_id` can be found with `account_id` in the sql db. We should avoid duplicates.

File: users/shared_account.go [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+	"github.com/trackit/trackit-server/models"
+	"github.com/trackit/trackit-server/mail"
+)
+
+// inviteUserRequest is the expected request body for the invite user route handler.
+type inviteUserRequest struct {
+	Email              string `json:"email" req:"nonzero"`
+	AccountId          int `json:"accountId"`
+	PermissionLevel    int `json:"permissionLevel"`
+}
```
Comment (thibautcornolti) [inline users/shared_account.go]: Bad indentation

File: users/shared_account.go [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+// addAccountToGuest adds an entry in shared_account table allowing a user
+// to share an access to all or part of his account
+func addAccountToGuest(ctx context.Context, db *sql.Tx, accountId int, permissionLevel int, guestId int, ownerId int) (error) {
+	const sqlstr = `INSERT INTO shared_account(
+			account_id, owner_id, user_id, user_permission, account_status
+		) VALUES (?, ?, ?, ?, ?)`
+	_, err := db.Exec(sqlstr, accountId, ownerId, guestId, permissionLevel, 0)
+	return err
+}
```
Comment (thibautcornolti) [inline users/shared_account.go]: Use Xo.

**Review by thibautcornolti — CHANGES_REQUESTED** [review]

**Review by simonmeyerrr — APPROVED** [review]

File: users/shared_account.go [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+			} else {
+				logger.Error("Error occured while creating new account for a guest.", err.Error())
+				return 403, errors.New("An error occured while inviting a new user. Please, try again.")
+			}
+		}
+	} else {
+		logger.Error("Error occured while checking if user already exist.", err.Error())
+		return 403, errors.New("An error occured while inviting a new user. Please, try again.")
+	}
+}
```
Comment (thibautcornolti) [inline users/shared_account.go]: I see a lot of errors, you should pack them in the top of the file
```
var (
   ...
)
```
(see `users/auth.go`)

**Review by thibautcornolti — CHANGES_REQUESTED** [review]

**Review by thibautcornolti — APPROVED** [review]

**Pull request merged**